docs/decisions/0015-serving-static-assets.rst

```diff
@@ -0,0 +1,125 @@
+
+**Serving assets should not *require* ASGI deployment.**
+  Deploying the LMS and Studio using ASGI would likely substantially improve the scalability of a Django-based streaming solution, but migrating and testing this new deployment type for the entire stack is a large task and is considered out of scope for this project.
+
+Decision
+--------
+
+URLs
+~~~~
+
+The format will be: ``https://{asset_server}/assets/apps/{app}/{learning_package_key}/{component_key}/{version}/{filepath}``
+
+The asset server will have a completely different domain from the LMS and Studio, and will not be a subdomain.
+
+A more concrete example: ``https://studio.assets.sandbox.openedx.io/apps/content_libraries/lib:Axim:200/xblock.v1:problem@826eb471-0db2-4943-b343-afa65a6fdeb5/v2/static/images/fig1.png``
+
+The ``version`` can be:
+
+* ``draft`` indicating the latest draft version (viewed by authors in Studio).
+* ``published`` indicating the latest published version (viewed by students in the LMS)
+* ``v{num}`` meaning a specific version–e.g. ``v20`` for version 20.
+
+Site Separation
+~~~~~~~~~~~~~~~
+
+The asset-serving site will run in the same process as the LMS or Studio process the links, so ``studio.assets.sandbox.openedx.io`` would point to the same place that ``studio.sandbox.openedx.org``. Being in the same process will make it easier to implement permissions checks and URL generation code.
+
+The switching between the two would happen with a new Middleware class that is implemented in openedx-learning. We would use this Middleware to make sure that assets are never served through the Studio and LMS URLs, and that no other Studio or LMS views are served through the asset server URL. This will require new configuration settings.
+
+View Implementation
+~~~~~~~~~~~~~~~~~~~
+
+The Learning Core will implement a Django REST Framework APIView that will provide the core functionality for serving assets. This view will be subclassed and customized by apps in LMS and Studio that want to make use of this functionality.
+
+The Learning Core will also provide auth related endpoints.
+
+Permissions
+~~~~~~~~~~~
+
+The Learning Core will not implement any permissions checking. It will be the responsibility of an app in Studio or the LMS add permissions classes to their subclass of the APIView.
+
+For instance, Studio would likely implement a simple permissions checking model that only examines the learning context and restricts access to course staff. LMS might eventually use a much more sophisticated model that looks at the individual Component that an asset belongs to.
+
+Cookie Authentication
+~~~~~~~~~~~~~~~~~~~~~
+
+Authentication will use a domain cookie on the root assets domain (e.g. ``*.assets.sandbox.openedx.io``). The cookie will have the ``Secure`` and ``HttpOnly`` attributes.
+
+OPEN QUESTIONS:
+
+* What's the best way to do handoff between LMS/Studio and get that cookie information over to the asset domains?
```

Comment:
One idea would be to have a one-time key that we store in cache and redirect the user's browser with it.  So it could be something like:

1. Browser makes a request to a new endpoint in Studio/LMS.
2. Studio/LMS generates a random token, and stores user information associated with it in our backend  cache (redis or memecached).
3. Studio/LMS then redirects the request to an auth URL on the static assets server with that random token in the querystring.
4. Asset server finds that entry in the cache, logs that user in, and removes the cache entry.
5. Asset server is now using session auth for that user.

It's cheating in a way because it relies on the asset server and the Studio or LMS instance being on the same service backend. But we're making that the case anyway because of permissions checking integration. And it's simpler than having signed requests (though it could ultimately be extended to go in that direction if we ever want to separate it into an entirely different service).